src/Sleep.js

```diff
@@ -3,77 +3,151 @@ class Sleep {
   }
+
+  // replaced
   calculateWeekSleepQuality(date, id, userRepo) {
     return userRepo.getWeekFromDate(date, id, this.sleepData).map((data) => `${data.date}: ${data.sleepQuality}`);
   }
+
   calculateAllUserSleepQuality() {
-    var totalSleepQuality = this.sleepData.reduce(function(sumSoFar, dataItem) {
-      sumSoFar += dataItem.sleepQuality;
+    let totalSleepQuality = this.sleepData.reduce((sumSoFar, sleepInstance) => {
+      sumSoFar += sleepInstance.sleepQuality;
       return sumSoFar;
-    }, 0)
-    return totalSleepQuality / this.sleepData.length
+    }, 0);
+    const averageSleepQuality = totalSleepQuality / this.sleepData.length;
+    return parseFloat(averageSleepQuality.toFixed(2)) || undefined;
+  }
+
+  getUsersWithQualityAbove3(endDate) {
+    const uniqueUsers = this.getUniqueUserIDs();
+    return uniqueUsers.filter((user) => {
+      let totalUserSleepQualityForWeek = this.getWeekOfData(user, endDate).reduce((sum, sleepInstance) => {
+        return (sum += sleepInstance.sleepQuality);
+      }, 0);
+      return totalUserSleepQualityForWeek / 7 > 3;
+    });
+  }
+  // Helper function for above
+  getUniqueUserIDs() {
+    return this.sleepData.reduce((users, sleepInstance) => {
+      if (!users.includes(sleepInstance.userID)) {
+        users.push(sleepInstance.userID);
+      }
+      return users;
+    }, []);
   }
+
+  //replaced with above to not use userRepo
   determineBestSleepers(date, userRepo) {
     let timeline = userRepo.chooseWeekDataForAllUsers(this.sleepData, date);
     let userSleepObject = userRepo.isolateUsernameAndRelevantData(this.sleepData, date, 'sleepQuality', timeline);
 
-    return Object.keys(userSleepObject).filter(function(key) {
-      return (userSleepObject[key].reduce(function(sumSoFar, sleepQualityValue) {
-        sumSoFar += sleepQualityValue
-        return sumSoFar;
-      }, 0) / userSleepObject[key].length) > 3
-    }).map(function(sleeper) {
-      return userRepo.getDataFromID(parseInt(sleeper)).name;
-    })
+    return Object.keys(userSleepObject)
+      .filter(function (key) {
+        return (
+          userSleepObject[key].reduce(function (sumSoFar, sleepQualityValue) {
+            sumSoFar += sleepQualityValue;
+            return sumSoFar;
+          }, 0) /
+            userSleepObject[key].length >
+          3
+        );
+      })
+      .map(function (sleeper) {
+        return userRepo.getDataFromID(parseInt(sleeper)).name;
+      });
+  }
+
+  getSleepWinnerForDay(date) {
+    const uniqueUsers = this.getUniqueUserIDs();
+    return uniqueUsers
+      .sort((userA, userB) => {
+        return this.calculateDailySleep(userB, date) - this.calculateDailySleep(userA, date);
+      })
+      .filter((user, index, sortedUniqueUsers) => {
+        return this.calculateDailySleep(user, date) === this.calculateDailySleep(sortedUniqueUsers[0], date);
+      }, []);
   }
+
+  // not sure where all below are being used, cannot find in scripts
```

Comment:
All these below can be removed if we don't find them being used?

Reply:
Yes, But I think The logic of them could be used if we try for an extension involving other winners.